src/lib/plugin.js

```
@@ -805,21 +807,46 @@ class FiveBellsLedger extends EventEmitter2 {
   }
 
   * _requestWithCredentials (options) {
-    const authToken = yield this._getAuthToken()
-    return yield request(Object.assign(
-      requestCredentials(this.credentials, authToken),
-      options))
+    // figure out which auth mechanism is supported
+    if (!this.supportedAuth) {
+      this.supportedAuth = yield this._getAuthMechanisms()
+    }
+
+    let requestCreds = null
+    if (this.supportedAuth === 'token') {
+      requestCreds = requestCredentials(this.credentials, yield this._getAuthToken())
+    } else if ((this.supportedAuth === 'basic')) {
+      requestCreds = requestCredentials(this.credentials)
+    }
+
+    return yield request(Object.assign(requestCreds, options))
+  }
+
+  * _getAuthMechanisms () {
+    let tryAuth = function * (options) {
+      let resp = yield request.post(this.ledgerContext.urls.message, options)
+      return +resp.statusCode
+    }.bind(this)
+
+    if ((yield tryAuth({auth: {bearer: 'invalidToken'}})) === 403) {
+      return 'token'
+    } else if ((yield tryAuth({auth: {user: 'abc', pass: 'abc'}})) === 403) {
```

Since there are only two options now you can just make this an `else` branch instead of `else if`. Also you shouldn't use fake usernames and passwords like this because then the plugin will behave incorrectly if an account happens to have these values as their credentials.

---

> Since there are only two options now you can just make this an else branch instead of else if

In fact, there are three options: 1) the ledger supports token auth, 2) the ledger supports basic auth, 3) something went wrong and we don't know what auth the ledger supports. 

The second if is necessary to ensure that the ledger supports basic auth. In other words, replacing the `else if` with a simple `else` would mean to implicitly assume the ledger supports basic auth if it does not support token auth.

>Also you shouldn't use fake usernames and passwords like this because then the plugin will behave incorrectly if an account happens to have these values as their credentials.

I changed it so that it sends a username that the ledger would not allow registering.

---

Right now Five Bells Ledgers are expected to support basic auth. You're changing it to also support token authentication. Practically that means there are only two options worth considering, so you should just check if it supports the new thing you're adding and if not fall back to what it was doing before. Theoretically it still supports using TLS client certs but I'm not sure we are really keeping that working

---

Concerning this code snippet, we can ignore other auth methods like client certs. It only has to decide if it should use a bearer token or basic auth (they are mutually exclusive).

If a (future/3rd party) ledger implementation supports neither bearer token nor basic auth, we will get a clear error message stating this fact. Doing this one additional check does not impose any noteworthy overhead (performance or otherwise), so why not keep it for the sake of clear error messages and "failing early".